### PR TITLE
[pipelining] Expose scoped stage forward context

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #196706
* #197488
* #196666
* #196665
* #196464
* #196463
* __->__ #196462
* #196956
* #196461
* #196460

## Background

Pipeline runtimes sometimes need the global logical stage and current
microbatch at the exact boundary where a stage module executes. A physical PP
rank is insufficient under virtual pipeline parallelism because one rank owns
multiple logical stages, while the microbatch identity belongs to runtime
execution rather than the model's input schema.

The previous revision transported this information through a reserved model
keyword and added an enable flag to every schedule. That made scheduler control
state part of the model signature, required kwargs copies and collision checks,
excluded traced stages, and coupled dynamic metadata inference to schedule
configuration.

## Design

`PipelineStageInfo` remains a frozen, slotted, keyword-only value containing the
global logical stage, step-local microbatch, and metadata-probe marker. A stage
now exposes `register_forward_context(factory)`, where the factory receives that
value and returns a fresh context manager around the stage computation.

The first version deliberately permits one live registration. A second
registration fails with an actionable error, and removing the returned
`RemovableHandle` permits replacement. This keeps ownership explicit and leaves
a compatibility-preserving path to multiple registrations if a concrete need
appears.

The default path performs only an empty-registry check: it constructs no info
object, copies no kwargs, and allocates no context manager. Registered runtime
forwards wrap the existing `forward_maybe_with_nosync` call. Dynamic metadata
inference wraps its representative `_compute_outputs` call with microbatch zero
and `is_metadata_inference=True`; static metadata setup does not execute the
factory. Manual and traced stages share the implementation through the stage
base class, so compiled and exported submodules retain their original graph
signatures.

Context teardown follows ordinary context-manager exception semantics. A
context is not allowed to suppress an exception raised by the stage module;
doing so produces an explicit runtime error instead of continuing without a
forward result.

## Boundaries

The registration is attached to `forward_one_chunk`, not schedule IR. Custom
actions receive it when they delegate to that method and otherwise own their
execution context. CUDA graph capture executes the Python context while
recording work, but replay does not re-enter Python; consumers must bind
replay-stable state during capture.

The implementation is built cleanly from the parent of the prior revision. It
adds no schedule constructor option, reserved keyword, schedule voting state,
or `_IR.py` behavior.

## Test Plan

```bash
python test/distributed/pipelining/test_stage.py PipelineStageForwardContextTest
python test/distributed/pipelining/test_stage.py PipelineStageMetadataInferenceTest
NCCL_DEBUG=WARN CUDA_VISIBLE_DEVICES=0,1,2,3 \
  python test/distributed/pipelining/test_stage.py
python test/distributed/pipelining/test_schedule.py
NCCL_DEBUG=WARN CUDA_VISIBLE_DEVICES=0,1,2,3 \
  python test/distributed/pipelining/test_schedule_multiproc.py
lintrunner --config=.lintrunner.toml \
  docs/source/distributed.pipelining.md \
  test/distributed/pipelining/test_schedule_multiproc.py \
  test/distributed/pipelining/test_stage.py \
  torch/distributed/pipelining/__init__.py \
  torch/distributed/pipelining/stage.py
sphinx-build -M html docs/source docs/build -j 8 -WT --keep-going
```

The Sphinx coverage report is 100% with zero undocumented Python objects. The
local full-tree post-check separately reports two pre-existing optional c10d
backend exports below this stack; the pipelining API introduced here is fully
documented.

AI assistance disclosure:
> This change was authored with assistance from an AI coding tool and reviewed by the author.

### CODE REVIEW

**Comment by sanketpurandare** [comment]
@claude provide a critical pr review, check for bugs, soundness of design, test coverage, redundant tests, slop code, cpmpactness of code, api design and provide feedback or suggest changes on how to make it better.

**Comment by sanketpurandare** [comment]
@claude Removed the unsupported traced `_PipelineStage`/FX plumbing and scoped the API to manual `PipelineStage`, whose forward boundary TorchTitan eager PP uses. The opt-in now requires complete static metadata, rejects dynamic inference because it has no real microbatch identity, stores the flag privately, excludes metadata from autograd inputs, and supports either direct keyword consumption or a standard `with_kwargs=True` pre-hook for strict module signatures. Tests now exercise `schedule.step()`, exact global stage/microbatch IDs, a real non-first same-rank stage handoff, the disabled default, collision rejection, and the static-metadata error. I retained explicit kwargs instead of ambient context because stage/microbatch identity is execution input state and must remain visible to tracing and custom forward contracts.

**Comment by sanketpurandare** [comment]
@claude Centralized the reserved keyword names and documented their global stage/microbatch semantics plus Python-integer compile specialization. Static capability and metadata requirement are now voted schedule-wide before metadata P2P, so every rank rejects an invalid mixed configuration consistently. Collision validation remains at dispatch because caller kwargs may differ by step, and the backward test proves metadata is excluded from saved autograd inputs. The focused stage tests and four-GPU rank-symmetric rejection test pass.

**Comment by sanketpurandare** [comment]
I used Codex to prepare this verification summary and reviewed it against the implementation and test results; it addresses the findings in this review: @claude Made the backward regression load-bearing by inspecting the saved autograd inputs, replaced the unreachable dynamic-stage branch with an invariant assertion, restored the compile/eval contract documentation, simplified the fixtures, and guarded the single-rank case explicitly. Collision validation deliberately remains at dispatch because pre-split caller kwargs may vary by microbatch and direct `forward_one_chunk()` calls must obey the same reserved-key contract. The five focused metadata tests, all 17 stage tests, and the four-GPU rank-symmetric rejection case pass.

File: torch/distributed/pipelining/stage.py [hunk 1]
```diff
@@ -1868,24 +1922,29 @@ def _warmup_forward_vote(
...
-            The accumulated product tensor after this stage's vote.
+            The two-element accumulated vote after this stage.
         """
-        my_vote = 0 if InferenceMode.needs_dynamic(self._user_meta, has_backward) else 1
-
-        my_vote_t = torch.tensor([my_vote], dtype=torch.int32, device=self.device)
+        supports_static = int(
+            not InferenceMode.needs_dynamic(self._user_meta, has_backward)
+        )
+        permits_dynamic = int(not self._pass_pipeline_metadata)
```
Comment (SherlockNoMad) [inline torch/distributed/pipelining/stage.py]: Me: What's metadata? 

Agent:  The word "metadata" is now badly overloaded. InferenceMode's own docstring reads "Pipeline-level metadata inference mode" — where "metadata" means tensor layout. This PR introduces
  pass_pipeline_metadata, PIPELINE_STAGE_INDEX_KEY, and a _PIPELINE_METADATA_KEYS frozenset where "pipeline metadata" means stage/microbatch identity, an unrelated concept, and routes it
  through the same vote tensor and the same PipeliningMetadataError. Two different meanings in adjacent lines of the same functions. I'll carry that into the review as a naming finding —
  something like pass_stage_context / PIPELINE_CONTEXT_* would keep them distinct.

**Comment by SherlockNoMad** [comment]
Me: How are pipeline_stage_index, pipeline_microbatch_index useful?
Agent: the PR never says, and nothing in the tree uses it.

 
I thought `pipeline_stage_index` should be constant for a stage, and should be available from Stage's property? 

I can imagine how `pipeline_microbatch_index` is useful... but still need to read other PR in the stack to understand how you plan to use it.

File: torch/distributed/pipelining/stage.py (hunk 1 above)
Comment (SherlockNoMad) [reply torch/distributed/pipelining/stage.py]: I am confused by why we need to pass `permits_dynamic` through the pipeline at all? 

Can `self.pass_pipeline_metadata` be a global config shared by all stages? do you see a case where different stage need a different `pass_pipeline_metadata` value?

**Review by SherlockNoMad — CHANGES_REQUESTED** [review]

File: docs/source/distributed.pipelining.md [hunk 2]
```diff
@@ -406,6 +406,62 @@ For example, `ScheduleGPipe` and `Schedule1F1B` are subclasses of `PipelineSched
 Whereas, `ScheduleInterleaved1F1B`, `ScheduleLoopedBFS`, `ScheduleInterleavedZeroBubble`, and `ScheduleZBVZeroBubble`
 are subclasses of `PipelineScheduleMulti`.
 
+### Accessing Stage Forward Information
```
Comment (SherlockNoMad) [inline docs/source/distributed.pipelining.md:409]: I didn't find example of how to access the current PipelineStageInfo from within the stage computation. 

I suppose it's in your DistMoE impl, but it's missing from the current PR.

**Review by SherlockNoMad — APPROVED** [review]

File: docs/source/distributed.pipelining.md (hunk 2 above)
Comment (sanketpurandare) [reply docs/source/distributed.pipelining.md:409]: You don't access it from stage computation, you set it outside via the registered context. DistMoE context or runtime does is not owned by a stage, module or layer. It is owned by the trainer runtime. So this extension allows the trainer to register a context manager that will be called around the stage forward providing it the stage id and microbatch id to chose the correct activation slot before the stage executes its forward

File: torch/distributed/pipelining/stage.py (hunk 1 above)
Comment (sanketpurandare) [reply torch/distributed/pipelining/stage.py]: Removed that design entirely. There is no permits_dynamic, pass_pipeline_metadata, reserved model kwarg, or metadata-vote input in the final implementation. A stage optionally registers one forward-context factory. Dynamic metadata inference invokes it with is_metadata_inference=True, static metadata setup does not execute the module, and real forwards receive their actual global stage and microbatch indices.

**Comment by sanketpurandare** [comment]
> Me: How are pipeline_stage_index, pipeline_microbatch_index useful? Agent: the PR never says, and nothing in the tree uses it.
> 
> I thought `pipeline_stage_index` should be constant for a stage, and should be available from Stage's property?
> 
> I can imagine how `pipeline_microbatch_index` is useful... but still need to read other PR in the stack to understand how you plan to use it.

Reworked the API around the concrete Dist-MoE use case. A physical PP rank may own several logical stages, while activation-memory and other runtime resources are selected for each (stage_index, microbatch_index) invocation. PipelineStageInfo gives a context factory one immutable, self-contained identity, which also lets one factory be registered across several stage objects. The stage index remains available on the stage itself; carrying it in the invocation value avoids coupling a shared consumer back to a particular stage instance. The public documentation now includes this use case and the API no longer changes the wrapped module's forward signature.
